### PR TITLE
[IMP] survey: make it possible to duplicate survey with badge

### PR DESCRIPTION
**Pull request by dharmrajsinh-jhala**
PURPOSE
Be able to duplicate a survey with a badge.

SPECIFICATION
If survey has a badge an try to duplicate it,
it should duplicate everything and remove the badge.
As badge is unique per certification, it's ok to let the user create a new badge for this new certification.

TaskID - 2265556
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/survey/views/survey_survey_views.xml [hunk 1]
```diff
@@ -121,7 +121,8 @@
                                             icon="fa-external-link"  target="_blank" class="btn-link pt-0"/>
                                     </div>
                                     <field name="certification_give_badge" attrs="{'invisible': ['|', ('certification', '=', False), ('users_login_required', '=', False)]}" />
-                                    <field name="certification_badge_id" attrs="{'invisible': ['|', ('certification_give_badge', '=', False), ('certification_badge_id', '!=', False)]}"
+                                    <field name="certification_badge_id"
+                                           attrs="{'invisible': ['|', ('certification_give_badge', '=', False), ('certification_badge_id', '!=', False)], 'required': [('certification_give_badge', '=', True), ('certification_badge_id', '=', False)]}"
```
Comment (dbeguin) [inline addons/survey/views/survey_survey_views.xml:125]: required : not sure ('certification_badge_id', '=', False) is necessary.
if give_badge = True, certification_badge_id should be set, and that's it.
unless I am missing something ?

**Comment by tde-banana-odoo** [comment]
Hello,

Sorry for time taken to review this one :) . I am not sure it is a good idea to relax constraint and their tests. We want to be sure that a certification granting a badge effectively has a badge set.

I see two solutions
  * duplicating the badge (like it seems you did for stable);
  * keeping your current PR but adding a required=True on view if badge giving mode is enabled. That way using the backend people cannot set a badge-certification without a badge configured;

First solution keeps the hard constraint, second one allow less strict models while still ensuring everything is enforced when using it in backend. If we go for second solution just check that if people somehow modified their survey (give badge but no badge configured) there is no crash in code (like skipping if badge is False).

What do you think @dbeguin ?

**Comment by dja-odoo** [comment]
@tde-banana-odoo I guess second one is already the case in this PR, isn't it ?

**Comment by tde-banana-odoo** [comment]
Did not see the required in the view :+1: 

Why did you choose this solution and not the duplicate of badge ? Just asking to know what was the reasoning behind this choice.

**Comment by tde-banana-odoo** [comment]
Hello,

I see places in the code where having a void certification_badge_id would lead to issues (notably _create_certification_badge_trigger). If there is no constraint at db-level, you have to protect core methods and raise a ValueError at the right level if someone tries to validate a certification without a badge. Remember people could simply call a write through RPC. Required in view is ok but not sufficient.

PS: I updated commit message, please keep it and don't rebase-force on it  :D 

Thanks,

**Comment by dja-odoo** [comment]
@tde-banana-odoo It was written in the task spec to let user create a new badge for new certification, that's why :) 

> As badge is unique per certification, it's ok to let the user create a new badge for this new certification.

**Comment by tde-banana-odoo** [comment]
Hello,

As it seems there is a misunderstanding, I just pushed a test showing how to create broken data in database. This is quite simple actually. By a write set certification_give_badge to True without giving a badge. A challenge is created, without reward. Then set certification_give_badge to False. Challenge is not removed as badge is used to remove unused challenges.

According to me relaxing constraint leads to potential database issues. Somehow either we don't create challenge if we have no reward_id, either we find another way to link challenges to surveys and ensure data is cleaned.

cc @dbeguin

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**